components/raftstore/src/store/fsm/peer.rs

```diff
@@ -5248,6 +5305,15 @@ where
             replicated_idx
         };
         assert!(compact_idx >= first_idx);
+
+        if let Some(idx) = witness_replicated_idx {
+            // do not compact the log not replicated to witness, except witness is the most
```

Comment:
I hava a question: 

I see that only one witness is considered here.

``` rust
let witness_peer = self.region().get_peers().iter().find(|p| p.is_witness);
```
 So if there are two witnesses, the first is the most lagged, but the second is not, should we also compact the log and why?

Reply:
nice catch